### ISSUE TITLE
Ability to type ahead in the drop down lists

### ISSUE THREAD

**Issue by antares1470** [constellation-app/constellation#347]
### Prerequisites

* [ ] Put an X between the brackets on this line if you have done all of the following:
    * Running the latest version of Constellation
	* Attached the ***Support Package*** via `Help` > `Support Package`
    * Checked the FAQs: https://github.com/constellation-app/constellation/wiki/FAQ
    * Checked that your issue isn't already filed: https://github.com/constellation-app/constellation/issues
    * Checked that there is not already a module that provides the described functionality: https://github.com/constellation-app/constellation/wiki/Catalogue-of-Repositories

### Description

Abilty to type ahead in the drop down lists

### Steps to Reproduce

1. [First Step]
2. [Second Step]
3. [and so on...]

**Expected behaviour:** Start typing in drop down list and it starts to list options based on what you have typed

**Actual behaviour:** nothing

**Reproduces how often:** n/a

### Additional Information

Any additional information, configuration or data that might be necessary to reproduce the issue.

**Comment by sol695510** [comment]
@antares1470 which drop down lists are you referring to, or are you saying to create general functionality that can be applied to all uses of drop down lists.

**Comment by antares1470** [comment]
> @antares1470 which drop down lists are you referring to, or are you saying to create general functionality that can be applied to all uses of drop down lists.

@sol695510 The drop down lists in case I believe refers to the single-choice and multi-choice parameters (which use drop down lists). In theory, the drop down lists can actually get quite large and can result in doing a lot of scrolling and searching without the ability to type ahead.

In theory I suppose this could be applied to more general combo boxes but I don't think it's necessary for this. @arcturus2 may be able to clarify this

**Comment by arcturus2** [comment]
@sol695510 and @antares1470 so single and multi choice options would be fine to start with. If this option could be extended to Text Fields that would be pretty cool though.

So for some background, we originally used the ControlsFX library for a lot of the GUI components but as Java started to build these in we switched it using them. JavaFX is open source now so maybe these features exist that we can use (https://openjfx.io/index.html)? Also, these "typeahead" options look like they are build into ControlsFX so perhaps we can check out they do it and switch the drop down lists to ControlxFX again.

**Comment by Auriga2** [comment]
Implemented on `SingleChoiceParameter`. `controlsfx.control.textfield.bindAutoCompletion` works only if the list contains simple strings, not for the `ComboBox<ParameterValue>`. Working with a custom function and a `StringConverter`, but trying to get it working with the `controlsfx.control.textfield.bindAutoCompletion`  and the 'StringConverter'

A bit tricky to implement on `MultiChoiceParameter`. Still working on it.

Also I am not sure how to extend this feature on `Text Fields`. What values do we filter here? Perhaps we can use `controlsfx.control.textfield.bindAutoCompletion` , but great if you can provide an example. For e.g., In the AttributeNodeProvider it has the attributes list, but it also has the filter implemented already. Looks  like it's based on the usage?

**Comment by Auriga2** [comment]
With the `TextFields.bindAutoCompletion`, when the user click on the Down arrow, it populates all items in the drop down, but when a value is selected (or when another value is selected when it already has a value), that is populated in the text box as expected, but it also displays it gain under it, as if when we type it. Not a big deal, but annoying.
![image](https://user-images.githubusercontent.com/64826928/111406892-fab90780-8722-11eb-9030-85537289b52b.png)

Also, tried even the below code, but it didn't populate the suggestions on certain `SingleChoiceParameter` drop downs, like `Type Category` in `Select Top N` plugin (always) and `Transaction Type` in `Split Nodes` plugin (Some times) in the Data Access View. Looks like this is because we set the dropdown values later, in the `updateParameters` (this ceases working occasionally, as with `Split Nodes` plugin) or using `params.addController`. Verified with the `Test Parameters` plugin- the `Robot Options` work with the pre populated values. But if I comment those lines (241-247), the values added in the `params.addController(REFRESH_PARAMETER_ID,...` does not auto populate when typing.
```
 field.setEditable(true);
        TextFields.bindAutoCompletion(field.getEditor(),
                field.getItems().stream().map(parameterValue ->
                        parameterValue.getObjectValue().toString()).collect(Collectors.toList()));
```



There are issues with the custom code, E.g.:
When the user click on the Down arrow, after selecting a value, it still display the sorted list, not the full list. If they need to see the full list again, the selected value should be cleared.  

But it might be a good idea to try to fix that, rather than changing different plugins.

**Comment by Auriga2** [comment]
Tried another option: `controlsfx.controls.SearchableComboBox` . This is easy to add, with a few lines of code, but in the `Select Top N` plugin it throws a `NoSuchElementException` exception when  changing the value in  'Mode' drop down, from 'Node' to 'Transaction'. Looks like this is because the number of items in the 'Type Category' drop down varies for 'Node' and 'Transaction' options in the 'Mode' parameter!

Could be because `params.addController(MODE_PARAMETER_ID...` is called twice.

**Comment by Auriga2** [comment]
The `NoSuchElementException` exception is not because `params.addController(MODE_PARAMETER_ID..`. is called twice. (I couldn't find a way to stop it, but added a check to prevent it processing further when the drop down list is not actually changed, which is better.)

So this is thrown only when the Drop down list changes from a longer list to a shorter list. (from 'Node' to 'Transaction')
It's thrown even if I try to clear the selection before changing drop down list, which is weird:
```
field.getSelectionModel().clearSelection();
field.setItems(FXCollections.observableList(options));
```
NoSuchElementException:
```
java.util.NoSuchElementException
	at java.base/java.util.AbstractList$Itr.next(AbstractList.java:377)
	at java.base/java.util.AbstractList.indexOf(AbstractList.java:188)	
	at javafx.controls/javafx.scene.control.skin.ComboBoxListViewSkin.getIndexOfComboBoxValueInItemsList(ComboBoxListViewSkin.java:467)
	at javafx.controls/javafx.scene.control.skin.ComboBoxListViewSkin.updateDisplayNode(ComboBoxListViewSkin.java:349)
	at javafx.controls/javafx.scene.control.skin.ComboBoxListViewSkin.getDisplayNode(ComboBoxListViewSkin.java:266)	
	at javafx.controls/javafx.scene.control.skin.ComboBoxBaseSkin.updateDisplayArea(ComboBoxBaseSkin.java:289)
	at javafx.controls/javafx.scene.control.skin.ComboBoxBaseSkin.lambda$new$7(ComboBoxBaseSkin.java:133)	
	at java.base/java.util.function.Consumer.lambda$andThen$0(Consumer.java:65)
	at javafx.controls/com.sun.javafx.scene.control.LambdaMultiplePropertyChangeListenerHandler.lambda$new$1(LambdaMultiplePropertyChangeListenerHandler.java:49)
	at javafx.base/javafx.beans.value.WeakChangeListener.changed(WeakChangeListener.java:86)
[catch] at javafx.base/com.sun.javafx.binding.ExpressionHelper$Generic.fireValueChangedEvent(ExpressionHelper.java:360)
	at javafx.base/com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:80)
	at javafx.base/javafx.beans.property.ObjectPropertyBase.fireValueChangedEvent(ObjectPropertyBase.java:106)
	at javafx.base/javafx.beans.property.ObjectPropertyBase.markInvalid(ObjectPropertyBase.java:113)
	at javafx.base/javafx.beans.property.ObjectPropertyBase.set(ObjectPropertyBase.java:147)
	at javafx.base/javafx.beans.property.ObjectProperty.setValue(ObjectProperty.java:72)
	at javafx.base/com.sun.javafx.binding.BidirectionalBinding$TypedGenericBidirectionalBinding.changed(BidirectionalBinding.java:601)
	at javafx.base/com.sun.javafx.binding.ExpressionHelper$Generic.fireValueChangedEvent(ExpressionHelper.java:360)
	at javafx.base/com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:80)
	at javafx.base/javafx.beans.property.ObjectPropertyBase.fireValueChangedEvent(ObjectPropertyBase.java:106)
	at javafx.base/javafx.beans.property.ObjectPropertyBase.markInvalid(ObjectPropertyBase.java:113)
	at javafx.base/javafx.beans.property.ObjectPropertyBase.set(ObjectPropertyBase.java:147)
	at javafx.controls/javafx.scene.control.ComboBoxBase.setValue(ComboBoxBase.java:151)
	at javafx.controls/javafx.scene.control.ComboBox.updateValue(ComboBox.java:520)
	at javafx.controls/javafx.scene.control.ComboBox.access$200(ComboBox.java:194)
	at javafx.controls/javafx.scene.control.ComboBox$3.changed(ComboBox.java:503)
	at javafx.base/com.sun.javafx.binding.ExpressionHelper$SingleChange.fireValueChangedEvent(ExpressionHelper.java:181)
	at javafx.base/com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:80)
	at javafx.base/javafx.beans.property.ReadOnlyObjectPropertyBase.fireValueChangedEvent(ReadOnlyObjectPropertyBase.java:74)
	at javafx.base/javafx.beans.property.ReadOnlyObjectWrapper.fireValueChangedEvent(ReadOnlyObjectWrapper.java:102)
	at javafx.base/javafx.beans.property.ObjectPropertyBase.markInvalid(ObjectPropertyBase.java:113)
	at javafx.base/javafx.beans.property.ObjectPropertyBase.set(ObjectPropertyBase.java:147)
	at javafx.controls/javafx.scene.control.SelectionModel.setSelectedItem(SelectionModel.java:105)
	at javafx.controls/javafx.scene.control.ComboBox$ComboBoxSelectionModel.lambda$new$0(ComboBox.java:560)
	at javafx.base/com.sun.javafx.binding.ExpressionHelper$SingleInvalidation.fireValueChangedEvent(ExpressionHelper.java:136)
	at javafx.base/com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:80)
	at javafx.base/javafx.beans.property.ReadOnlyIntegerPropertyBase.fireValueChangedEvent(ReadOnlyIntegerPropertyBase.java:72)
	at javafx.base/javafx.beans.property.ReadOnlyIntegerWrapper.fireValueChangedEvent(ReadOnlyIntegerWrapper.java:102)
	at javafx.base/javafx.beans.property.IntegerPropertyBase.markInvalid(IntegerPropertyBase.java:114)
	at javafx.base/javafx.beans.property.IntegerPropertyBase.set(IntegerPropertyBase.java:148)
	at javafx.controls/javafx.scene.control.SelectionModel.setSelectedIndex(SelectionModel.java:69)
	at javafx.controls/javafx.scene.control.ComboBox$ComboBoxSelectionModel.updateItemsObserver(ComboBox.java:656)
	at javafx.controls/javafx.scene.control.ComboBox$ComboBoxSelectionModel.access$500(ComboBox.java:539)
	at javafx.controls/javafx.scene.control.ComboBox$ComboBoxSelectionModel$1.invalidated(ComboBox.java:577)
	at javafx.base/javafx.beans.WeakInvalidationListener.invalidated(WeakInvalidationListener.java:83)
	at javafx.base/com.sun.javafx.binding.ExpressionHelper$Generic.fireValueChangedEvent(ExpressionHelper.java:348)
	at javafx.base/com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:80)
	at javafx.base/javafx.beans.property.ObjectPropertyBase.fireValueChangedEvent(ObjectPropertyBase.java:106)
	at javafx.base/javafx.beans.property.ObjectPropertyBase.markInvalid(ObjectPropertyBase.java:113)
	at javafx.base/javafx.beans.property.ObjectPropertyBase.set(ObjectPropertyBase.java:147)
	at javafx.controls/javafx.scene.control.ComboBox.setItems(ComboBox.java:310)
	
	at au.gov.asd.tac.constellation.plugins.gui.SingleChoiceInputPane.lambda$new$3(SingleChoiceInputPane.java:138)
	at javafx.graphics/com.sun.javafx.application.PlatformImpl.lambda$runLater$10(PlatformImpl.java:428)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at javafx.graphics/com.sun.javafx.application.PlatformImpl.lambda$runLater$11(PlatformImpl.java:427)
	at javafx.graphics/com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:96)
	at javafx.graphics/com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
	at javafx.graphics/com.sun.glass.ui.win.WinApplication.lambda$runLoop$3(WinApplication.java:174)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by Auriga2** [comment]
The same question is raised in https://github.com/controlsfx/controlsfx/issues/1320 but apparently that lib is unmaintained and they have closed the ticket! So may be we should move forward with other options.

**Comment by abhinayagarwal** [comment]
The library is maintained. The last release 11.1.0 was done 27 days ago. Please open a new issue and we will have a look at it.

**Comment by Auriga2** [comment]
Fixed the SingleChoiceParameterType drop downs locally, and created a PR. So we are no longer waiting on https://github.com/controlsfx/controlsfx/issues/1320
But MultiChoiceParameterType requires more work and there's a ticket in controlsfx backlog https://github.com/controlsfx/controlsfx/issues/1160

**Comment by Auriga2** [comment]
Created https://github.com/constellation-app/constellation/issues/1012 to handle multi choice options seperately.
